### PR TITLE
Ability to disable the guard bytes and poison of heap blocks to allow…

### PR DESCRIPTION
**Pull request by StephenHillAtViaivi**
… dynamic code analysis tools, such as Boundschecker, to detect read-overruns.

### CODE REVIEW

**Comment by StephenHillAtViaivi** [comment]
Hmmm, this changeset doesn't fix the problem with new() as the accounting information is added on the end (which is rather a shame).

**Comment by basvodde** [comment]
What was the purpose?

**Comment by StephenHillAtViaivi** [comment]
Currently CppuTest allocates extra trailing bytes on the end of any buffer dynamically allocated and fills them with a pattern. This pattern is checked when de-allocated, which is great for detecting write-overruns, but cannot detect read-overruns. There are many tools (Boundschecker, AppVerifier, Valgrind) which can detect read and write-overruns, but they are defeated by these trailing bytes. This change-set is designed to allow these trailing bytes to be disabled in the case where these tools are used. I will be updating this changeset with some fixes later.

**Comment by basvodde** [comment]
Ah ok. That makes sense.

**Comment by StephenHillAtViaivi** [comment]
Add fixes for plugins

File: src/CppUTest/MemoryLeakDetector.cpp [hunk 1]
```diff
@@ -513,21 +519,25 @@ char* MemoryLeakDetector::reallocateMemoryAndLeakInformation(TestMemoryAllocator
 
 void MemoryLeakDetector::invalidateMemory(char* memory)
 {
+#ifndef CPPUTEST_DISABLE_HEAP_POISON
```
Comment (basvodde) [inline src/CppUTest/MemoryLeakDetector.cpp]: Why is this @ifdef needed?

File: include/CppUTest/MemoryLeakDetector.h [hunk 2]
```diff
@@ -216,7 +216,11 @@ class MemoryLeakDetector
     void removeMemoryLeakInformationWithoutCheckingOrDeallocatingTheMemoryButDeallocatingTheAccountInformation(TestMemoryAllocator* allocator, void* memory, bool allocatNodesSeperately);
     enum
     {
+#ifdef CPPUTEST_DISABLE_MEM_CORRUPTION_CHECK
+        memory_corruption_buffer_size = 0
+#else
         memory_corruption_buffer_size = 3
```
Comment (basvodde) [inline include/CppUTest/MemoryLeakDetector.h]: I guess here you also want to do the sizeof(guardbyte) ?

Comment (basvodde) [reply include/CppUTest/MemoryLeakDetector.h]: I guess by making GuardBytes empty when disabled, you do not require this variable at all and most of the #ifdef can be removed?

File: src/CppUTest/MemoryLeakDetector.cpp [hunk 3]
```diff
@@ -481,7 +483,11 @@ SimpleMutex *MemoryLeakDetector::getMutex()
 
 static size_t calculateVoidPointerAlignedSize(size_t size)
 {
+#ifndef CPPUTEST_DISABLE_MEM_CORRUPTION_CHECK
```
Comment (basvodde) [inline src/CppUTest/MemoryLeakDetector.cpp]: Why this one?

File: src/CppUTest/MemoryLeakDetector.cpp (hunk 1 above)
Comment (StephenHillAtViaivi) [reply src/CppUTest/MemoryLeakDetector.cpp]: Address Sanitizer, Boundschecker and Appverifier poison all buffers with a pattern, so this is just unnecessary. I have made it a different conditional as some dynamic code analysis tools might not provide this feature (e.g. Valgrind).

File: include/CppUTest/MemoryLeakDetector.h (hunk 2 above)
Comment (StephenHillAtViaivi) [reply include/CppUTest/MemoryLeakDetector.h]: Ah, I can't do this as the Microsoft tool-chain doesn't allow zero length arrays.

**Comment by basvodde** [comment]
I'm uncomfortable with the #ifdef in the code. We are trying to remove #if and not add more. Still, I'll merge this as I understand the reason and cannot (yet) think of a better solution.

Although... just speculating... a different solution might be a command line option to turn it on/off. Most of it could be done dynamically. Unfortunately that means it would need to be off outside the tests so that it can be decided when parsing the parameters. Then, for global objects, the memory corruption check isn't done... that might lead to problems.. especially considering the nodes allocated differently...

Merged for now but requires more thought.

**Pull request merged**